RFD.md

```diff
@@ -0,0 +1,217 @@
+
+* In-flight content protection will be implemented by encrypting all traffic.
+* On-disk logs protection will be limited to file permissions (0600).
+* Token-based authorization will be used to control who can see output from which commands (see the dedicated section above).
+
+In a real production scenario we may want to encrypt on-disk logs (either do it ourselves or provide the operator with guidelines on how to achieve it via `dm-crypt`, encrypted EBS, or other technologies).
+
+#### Container isolation
+
+Since the system will be used to execute arbitrary commands from the users, we need to ensure full isolation of the user command from the underlying host OS. We're going to use Linux kernel namespaces to achieve isolation along the following dimensions:
+
+* PID namespace - isolates user commands from the host OS process namespace to avoid leaking process information from the host into the container.
+
+* Network namespace - isolates container networking stack from that of the host OS.
+
+* Mount namespace - allows us to fully isolate command's filesystem from that of the host OS (and protect host OS `/proc` from being accessed by the user command to gain access to any privileged information on the host).
+
+##### Scope limits
+
+For this exercise, we're going to accept the following constraints:
+
+* Networking namespace will be used to separate the user command from the host OS network, but we're not going to build any bridges, etc to connect container network to the host.
+
+* We're going to create a new clean filesystem by extracting an Alpine Root FS image per command, which will use 5Mb of disk space per container. In a production environment we would probably rely on some kind of layered filesystem like overlayfs to save on disk resources.
+
+#### Availability and resource limits
+
+In a multi-tenant environment like the system in question, we need to ensure proper resource limits to reduce the potential impact of noisy or abusive clients on the underlying host OS.
+
+We're going to use linux Cgroups to apply the following limits on each command:
+
+* Memory usage – a static per-command limit will be applied on each container.
+
+* CPU usage - each command will be limited to a small constant amount of CPU resources, preventing it from negative affecting the host OS and other commands by consuming too much CPU.
+
+* Disk IO - each command will be limited by the amount of disk IO it could perform to limit the effects from a single command's high IO load on the host OS and other commands.
+
+##### Scope limits
+
+All limits will be statically hardcoded within the server, while in a production system we'd probably have some kind of configuration file for default limits and would expand the API to allow users to specify their own limits for each command (like Docker does).
+
+In a production environment, we may want to introduce per-user resource limits, API call rate limits, etc to further protect the system from abuse.
+
+### API
```

Comment:
Could you also include an example of what the API of the library might look like?

Reply:
I've added a section with a proposed API design, but in my opinion, it is an internal implementation detail and I do not believe it belongs in a design document unless we plan to make that API public.

Reply:
I disagree - the challenge has 3 parts:

1. a library with an API for working with jobs
2. a GRPC server that uses the API of the library
3. a GRPC client

In this sense, the library API *is* public, and a well designed API is important in order to ensure that you can cleanly integrate the library functionality into your GRPC server.

Reply:
If we look at it from that point of view, I agree - it makes sense. I personally would prefer to let the highest level API dictate the implementation and even the design of lower levels to avoid designing in a vacuum. Basically, I'd build the implementation of the top level API and then "extract" wherever I use on the lowest level into a low level API (see what Docker did with Moby as a good example) instead of designing upfront it on two levels and aiming to align them together when I'm writing the code. 

But I definitely see your point to this being a part of the challenge itself and will update the doc to make sure the APIs are documented before the implementation.